paymentgateway/paypal/classes/ipn.php

```diff
@@ -0,0 +1,185 @@
+            } else {
+                $data->$property = null;
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * Reads all data from an IPN. Extracts data from it and creates $req for later use.
+     *
+     * @param object $post The IPN POST request.
+     * @return object $data Data from the IPN that has been processed.
+     * @throws \moodle_exception
+     */
+    public function process_ipn($post) {
+
+        $this->req = 'cmd=_notify-validate';
+
+        $postdata = new \stdClass();
+        foreach ($post as $key => $value) {
+                $this->req .= "&$key=".urlencode($value);
+                $postdata->$key = fix_utf8($value);
+        }
+
+        if (empty($postdata->custom)) {
+            throw new \moodle_exception('invalidrequest', 'core_error', '', null, 'Missing request param: custom');
+        }
+
+        $custom = explode('-', $postdata->custom);
+
+        if (empty($custom) || count($custom) < 2) {
+            throw new \moodle_exception('invalidrequest', 'core_error', '', null, 'Invalid value of the request param: custom');
+        }
+
+        $data = $this->set_data($postdata);
+
+        $data->userid = (int)$custom[0];
+        $data->courseid = (int)$custom[1];
+
+        return $data;
+    }
+
+    /**
+     * Validates the transaction data.
+     * 
+     * @param \stdclass $data The transaction data.
+     */
+    public function validate($data) {
+        global $CFG;
+
+        // Open a connection back to PayPal to validate the data.
+        $paypaladdr = empty($CFG->usepaypalsandbox) ? 'ipnpb.paypal.com' : 'ipnpb.sandbox.paypal.com';
+        $c = new \curl();
+        $options = array(
+            'returntransfer' => true,
+            'httpheader' => array('application/x-www-form-urlencoded', "Host: $paypaladdr"),
+            'timeout' => 30,
+            'CURLOPT_HTTP_VERSION' => CURL_HTTP_VERSION_1_1,
+        );
+        $location = "https://$paypaladdr/cgi-bin/webscr";
+        $result = $c->post($location, $this->req, $options);
+
+        if ($c->get_errno()) {
+            throw new \moodle_exception('errpaypalconnect', 'enrol_paypal', '', array('url' => $paypaladdr, 'result' => $result),
+                json_encode($data));
+        }
+
+        return $result;
+    }
+
+    /**
+     * Checks if anything is wrong with transaction data, and deals
+     * with errors by adding them to error_info.
+     *
+     * @param object $data
+     * @return bool $noerror
+     */
+    private function is_ipn_data_correct(&$data) {
+        global $DB;
+
+        $noerror = true;
+
+        $currency = get_config('paymentgateway_paypal', 'currency');
+        $cost = $DB->get_field('tool_paymentplugin_course', 'cost', array('courseid' => $data->courseid));
+
+        // Check that course price and currency matches.
+        $error_info = "";
+        if ($data->mc_currency != $currency) {
+            $error = false;
```

Comment:
This should be $noerror, currently no error would ever be handled correctly, as noerror never changes from true. A better pattern here would be to remove all the booleans completely, and instead check !empty($error_info) {handle error}